### PR TITLE
AutoMigration: Fix sb10 migration when main config contains `require`

### PR DESCRIPTION
**Pull request by ndelangen**
## What I did

This ensures 2 things:

1. the upgrade command does not fail when trying to load an faux-ESM file
   The work around is: we create a temp copy of the main config file, prepend a compat-banner, load and then delete it.
2. add an auto-migration to turn the faux ESM file into valid ESM-file by adding a header adding a reference to `require`

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

- Have a sandbox or project on SB8 (I've used mealdrop myself)
- add a `require.resolve` for any reason. But not in the getAbsolutePath way. (I wrapped a single addon)
- Upgrade to Sb10 via the upgrade command.
- The expectation is that `main.ts` gets modified to ensure a `const` of `require` exists.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-32558-sha-aadcab7f`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-32558-sha-aadcab7f sandbox` or in an existing project with `npx storybook@0.0.0-pr-32558-sha-aadcab7f upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-32558-sha-aadcab7f`](https://npmjs.com/package/storybook/v/0.0.0-pr-32558-sha-aadcab7f) |
| **Triggered by** | @ndelangen |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`norbert/sb10-upgrade-with-require-in-main`](https://github.com/storybookjs/storybook/tree/norbert/sb10-upgrade-with-require-in-main) |
| **Commit** | [`aadcab7f`](https://github.com/storybookjs/storybook/commit/aadcab7fd9af54c8b2457ea5c02eb0607d7c9436) |
| **Datetime** | Thu Sep 25 13:47:52 UTC 2025 (`1758808072`) |
| **Workflow run** | [18009620740](https://github.com/storybookjs/storybook/actions/runs/18009620740) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=32558`_
</details>

## Summary by CodeRabbit

* **New Features**
  * CLI automigrate can detect ESM main configs that use legacy require and offer/apply a compatibility banner to restore compatibility.

* **Bug Fixes**
  * Added a temporary compatibility fallback to avoid startup failures when `require` is not available while loading the main config.

* **Tests**
  * Added tests for detection, banner insertion, dry-run behavior, and edge cases.

### CODE REVIEW

File: code/core/src/common/utils/load-main-config.ts [hunk 1]
```diff
@@ -25,6 +29,33 @@ export async function loadMainConfig({
     if (!(e instanceof Error)) {
       throw e;
     }
+    if (e.message.includes('require is not defined')) {
+      logger.info(
+        'loading main config failed, trying a temporary fix, please ensure the main config is valid ESM'
```
Comment (valentinpalkovic) [inline code/core/src/common/utils/load-main-config.ts]:
```suggestion
        'Loading main config failed, trying a temporary fix. Please ensure the main config is valid ESM'
```

File: code/core/src/common/utils/load-main-config.ts [hunk 2]
```diff
@@ -25,6 +29,33 @@ export async function loadMainConfig({
...
+      );
+      const comment =
+        '// end of storybook 10 migration assistant header, you can delete the above code';
```
Comment (valentinpalkovic) [inline code/core/src/common/utils/load-main-config.ts]:
```suggestion
        '// end of Storybook 10 migration assistant header, you can delete the above code';
```

File: code/lib/cli-storybook/src/automigrate/helpers/mainConfigFile.ts [hunk 3]
```diff
@@ -228,3 +228,49 @@ export const updateMainConfig = async (
...
+export function containsRequireUsage(content: string): boolean {
+  // Check for require() calls
+  const requireCallRegex = /\brequire\(/;
+  const requireDotRegex = /\brequire\./;
+  return requireCallRegex.test(content) || requireDotRegex.test(content);
+}
+
+/** Check if the file already has the require banner */
+export const bannerComment =
+  '// end of storybook 10 migration assistant header, you can delete the above code';
```
Comment (valentinpalkovic) [inline code/lib/cli-storybook/src/automigrate/helpers/mainConfigFile.ts]:
```suggestion
  '// end of Storybook 10 migration assistant header. You can delete the above code';
```

File: code/lib/cli-storybook/src/automigrate/fixes/fix-faux-esm-require.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    // Check if the file contains require usage
+    if (!isWithRequire) {
+      return null;
+    }
+
+    return { storybookVersion };
+  },
+
+  prompt() {
+    return dedent`We have detected that your main config file is in ESM format but contains 'require' references. This will cause issues in Storybook 10. We'll add a compatibility banner to make it work.`;
```
Comment (valentinpalkovic) [inline code/lib/cli-storybook/src/automigrate/fixes/fix-faux-esm-require.ts]:
```suggestion
    return dedent`Main config is ESM but uses 'require'. This will break in Storybook 10; Adding compatibility banner`;
```

Making the prompt less verbose, since in the new upgrade flow, the prompt is shown as a checkbox item, and we want to keep the prompt as short as possible.

File: code/lib/cli-storybook/src/automigrate/fixes/fix-faux-esm-require.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    if (isWithBanner) {
+      return null;
+    }
+
+    // Check if the file contains require usage
+    if (!isWithRequire) {
+      return null;
+    }
+
+    return { storybookVersion };
```
Comment (valentinpalkovic) [inline code/lib/cli-storybook/src/automigrate/fixes/fix-faux-esm-require.ts]: Returning `true` is enough, since `storybookVersion` isn't really used in the `run` function.

**Review by valentinpalkovic — COMMENTED** [review]
Great job! Added some minor comments.

**Comment by ndelangen** [comment]
@yannbf I QA'd this on the chromatic monorepo, and it ran, but due to the fact, that `dirname` existed twice after the added header, it broke.

My idea is that we might be able to update/change the header that gets injected.
We can avoid exposing the `dirname`, by maybe importing like so:
```ts
import * as requireCompatPath from 'node:path';
```
